src/api/v1/users/register.ts

```diff
@@ -0,0 +1,153 @@
+import * as QRCode from '../../../services/qrcode';
+import * as PdfService from '../../../services/pdf';
+import { EmailTemplate } from '../../../services/email/templates';
+import { RoleEnum } from '../../../auth/types';
+import Roles from '../../../auth/roles';
+
+
+interface RegisterRequest extends Hapi.Request {
+  payload: {
+    organisationId: number
+    name: string
+    gender: string
+    birthYear: number
+    email: string
+    phoneNumber: string
+    postCode: string
+    emailConsent: boolean
+    smsConsent: boolean
+  };
+}
+
+
+export default [
+  {
+    method: 'POST',
+    path: '/users/register/visitor',
+    options: {
+      auth: false,
+      validate: {
+        payload: {
+          organisationId: id.required(),
+          name: visitorName.required(),
+          gender: gender.required(),
+          birthYear: birthYear.required(),
+          email: email.required(),
+          phoneNumber: phoneNumber.allow(''),
+          postCode: postCode.allow(''),
+          emailConsent: emailConsent.default(false),
+          smsConsent: smsConsent.default(false),
+        },
+      },
+      response: { schema: response },
+    },
+    handler: async (request: RegisterRequest, h: Hapi.ResponseToolkit) => {
+      const { knex, payload, server: { app: { EmailService } } } = request;
+
+      /*
+       * Preliminaries
+       * (Can possibly eventually be removed into pre-requisites)
+       */
+      // Check user doesn't already exist
+      if (await Visitors.exists(knex, { where: { email: payload.email } })) {
+        throw Boom.conflict('User with this e-mail already registered');
+      }
+      if (!(await CommunityBusinesses.exists(knex, { where: { id: payload.organisationId } }))) {
+        throw Boom.badRequest('Unrecognised organisation');
+      }
+
+      // Fetch the community business admin
+      const cb = await CommunityBusinesses.getOne(knex, {
+        where: { id: payload.organisationId, deletedAt: null },
+      });
+      const [admin] = await CbAdmins.fromOrganisation(knex, { id: payload.organisationId });
+
+      if (!admin) {
+        throw Boom.badData('No associated admin for this organisation');
+      }
+
+      /*
+       * Main actions
+       */
+      // Create the visitor
+      // // TODO: This should actually be part of the Visitor model
+      // //       However, this breaks the typings and it's not obvious
+      // //       right now how to accomodate it.
+      const visitor = await knex.transaction(async (trx) => {
+        const visitor = await Visitors.add(trx, Visitors.create(payload));
+        await Roles.add(trx, {
+          role: RoleEnum.VISITOR,
+          userId: visitor.id,
+          organisationId: payload.organisationId,
+        });
+        return visitor;
+      });
+
+      // Send QR code to visitor
+      // // generate QR code data URL
+      const qrCode = await QRCode.create(visitor.qrCode);
```

Comment:
Can this be done inside the transaction? If an error is thrown then the creation of the user will be rolled back.

Reply:
Possibly. I'm not sure what the deal is with doing non-DB operations inside a transaction. I'll look into it.

Reply:
Although in fact, really the only critical ops for this endpoint end at line 102. If the transaction completes successfully we are happy, since the rest of the operations can be repeated without much bother (e.g. the CB admin can request that the code is e-mailed to the visitor again).

Maybe if the latter half of the handler throws at any stage, we can return a [202](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#2xx_Success) response, which doesn't quite fit, but is the closest I can find.

Reply:
Your latter suggestion will involve changes to the client so that there is notification of the error. Is there a corresponding `twine-visitor` PR?

My preference would be for new users to not be able to sign up unless their qrcode has been sent, otherwise you end up with the edge case where they're not allowed to sign up again because of email constraints and have to talk to an admin to get their code.